### ISSUE TITLE
Deprecate evaluated arguments

### ISSUE THREAD

**Issue by theofidry** [nelmio/alice#498]
While porting the tests from v2 to v3, I've encountered the following test:

```php
public function testLoadParsesFakerDataWithPhpArgs()
    {
        $res = $this->loadData([
            self::USER => [
                'user0' => [
                    'username' => '<dateTimeBetween("yest"."erday", strrev("omot")."rrow")>',
                ],
            ],
        ]);

        $this->assertInstanceOf('DateTime', $res['user0']->username);
        $this->assertGreaterThanOrEqual(strtotime("yesterday"), $res['user0']->username->getTimestamp());
        $this->assertLessThanOrEqual(strtotime("tomorrow"), $res['user0']->username->getTimestamp());
    }
```

This test is giving me a lot of trouble: it is hard to image how the string `"yest"."erday", strrev("omot")."rrow"'` show be magically evaluated in `"yesterday", "tomorrow"'.

This is "doable" only if each function argument is evaluated. However the strategy in v3 is slightly different: we have a lexer and a parser to be able to figure out how to evaluate a string. For example:

`'<dateTimeBetween("yesterday", "tomorrow")>'` will result in:

```php
new FunctionCallValue('dateTimeBetween', ['yesterday', 'tomorrow']);
```

This allows more complex and safer combination, for example nested calls: `'<f(<g($baz)>, $arg)>'`. As of now, it still has some limitations: `'<f(<g($baz, $faz)>, $arg)>'` will fail because cannot guess that the first comma is for the nested function and not for the parent function. I think I can manage to solve that somehow, and as a result, an alternative in v3 for `'<dateTimeBetween("yest"."erday", strrev("omot")."rrow")>'` would be `'<dateTimeBetween(<("yest"."erday")>, <(strrev("omot")."rrow")>)>'`.

@tshelburne does this looks like an acceptable solution to you?

**Comment by tshelburne** [comment]
I think that works, though it introduces a lot of extra complexity - it feels like those kinds of specific needs should really be handled with custom providers or custom parsers (if we are allowing registration of custom parsers). I'm not against the approach if you want to support it, but it seems like overdoing things to me. I imagine there are use-cases out there, but IMO, that's too much logic in a fixture file.

**Comment by theofidry** [comment]
>it feels like those kinds of specific needs should really be handled with custom providers or custom parsers

I feel the same, but well that was a supported feature in v2...

>if we are allowing registration of custom parsers

Easier than ever ;)

>I'm not against the approach if you want to support it, but it seems like overdoing things to me. I imagine there are use-cases out there, but IMO, that's too much logic in a fixture file.

Then are you ok to drop it for now and decide on what to do with this one after the first beta? It will either be: A) I find an easy way to support things like `'<f(<g($baz, $faz)>, $arg)>'` in which case the migration path is really easy, B) I don't find one and this feature will be completely deprecated in favour of a custom provider

**Comment by tshelburne** [comment]
I think it makes sense either way - perhaps hammer on it a bit to see if you get it working quickly, and if not, just create a ticket in GH to consider adding it back in where users can voice their opinions.

**Comment by theofidry** [comment]
Nested function support has been added in 3.x, so the correct way to write `'<dateTimeBetween("yest"."erday", strrev("omot")."rrow")>'` in 3.x is `'<dateTimeBetween(<("yest"."erday")>, <(strrev("omot")."rrow")>)>'`.

The only part left then is deprecating this behaviour in 2.x but I'm honestly lost at how to do that... It might be better to just document it. The user will have an exception right away in 3.x anyway.

**Comment by tshelburne** [comment]
Yea, just add it to the upgrade notes.

**Comment by theofidry** [comment]
Closed via #563
